examples/extensible-interface-files/core-plugins/basic-example/Example.hs

```diff
@@ -0,0 +1,245 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Example where
+
+import Bag
+import GHC.Plugins
+import GHC.Iface.Load hiding (loadCore)
+import GHC.Iface.Syntax
+import GHC.IfaceToCore hiding (tcIfaceModGuts)
+import GHC.Tc.Types
+import GHC.Tc.Utils.Monad
+import GHC.Hs.Binds
+import GHC.Hs.Extension
+import GHC.Core.ConLike
+import Maybes as M
+import GHC.Iface.Env
+import GHC.Iface.Binary
+import GHC.Core.InstEnv
+import GHC.Core.FamInstEnv
+import GHC.CoreToIface
+
+
+-- We use the `interfaceWriteAction` to access the core bindings that are in the
+-- `ModGuts` as a result of the whole core pipeline. We then use serialise those
+-- bindings into the interface file using `registerInterfaceDataWith` to produce
+-- an extensible interface field from a plugin.
+
+-- We use the `typeCheckResultAction` to intercept the start of the core pipeline
+-- and perform our inlining step before any of the other core steps have run,
+-- including GHC's normal optimisation passes.
+plugin :: Plugin
+plugin = defaultPlugin { typeCheckResultAction = corePlugin
+                       , interfaceWriteAction  = interfacePlugin
+                       }
+
+-- Here we perform the serialisation of the `mg_binds` field from the `ModGuts`.
+-- Note, the deserialisation loading functions later require the `SrcSpan`, so
+-- we include it in the serialised data.
+interfacePlugin :: [CommandLineOption] -> HscEnv -> ModDetails -> ModGuts -> PartialModIface -> IO PartialModIface
+interfacePlugin _ env _ guts iface = do
+  registerInterfaceDataWith "plutus/core-bindings" env $ \bh ->
+    putWithUserData (const $ return ()) bh (mg_loc guts, map toIfaceBind' $ mg_binds guts)
+  return iface
+
+
+isRealBinding :: Bind Id -> Bool
+isRealBinding (NonRec n _) = isExternalName (idName n)
+isRealBinding _ = True
+
+
+toIfaceBind' :: Bind Id -> (Bool, IfaceBinding)
+toIfaceBind' b = (isRealBinding b, toIfaceBind b)

```

Comment:
```suggestion
  where
    toIfaceBind' :: Bind Id -> (Bool, IfaceBinding)
    toIfaceBind' b = (isRealBinding b, toIfaceBind b)

    where
      -- | We are only interested in ... bindings, that is excluding internal bindings
      -- ...???
      isRealBinding :: Bind Id -> Bool
      isRealBinding (NonRec n _) = isExternalName (idName n)
      isRealBinding _ = True

```
Can use `where` here as it's local to `interfacePlugin` only and not used anywhere else.

Also: What does Real mean?